lib/scholar/decomposition/pca.ex

```diff
@@ -0,0 +1,220 @@
+      value of `:num_features` and `:num_samples` if `num_components` is `nil`.
+
+    * `:num_features` - Number of features in the training data.
+
+    * `:num_samples` - Number of samples in the training data.
+
+    * `:decomposer` - Matrix used in actual decomposition. It is the unitary matrix from SVD factorization.
+  """
+  deftransform fit(x, opts \\ []) do
+    opts = NimbleOptions.validate!(opts, @fit_opts_schema)
+
+    if Nx.rank(x) != 2 do
+      raise ArgumentError,
+            "expected input to have shape {n_samples, n_features}, got tensor with shape: #{inspect(Nx.shape(x))}"
+    end
+
+    num_components =
+      case opts[:num_components] do
+        :none ->
+          {num_samples, num_features} = Nx.shape(x)
+          Kernel.min(num_features, num_samples)
+
+        %Nx.Tensor{} = num ->
+          num
+
+        num when is_number(num) ->
+          num
+
+        num ->
+          raise ArgumentError,
+                ":num_components must be :none, a number, or a tensor, got: #{inspect(num)}"
+      end
+
+    fit_n(x, num_components)
+  end
+
+  defnp fit_n(x, num_components) do
+    {num_samples, num_features} = Nx.shape(x)
+    mean = Nx.mean(x, axes: [0])
+    x = x - mean
+    {decomposer, singular_values, components} = Nx.LinAlg.svd(x)
+    explained_variance = singular_values * singular_values / (num_samples - 1)
+    explained_variance_ratio = explained_variance / Nx.sum(explained_variance)
+
+    type = Nx.type(num_components)
+    clipped = Nx.clip(num_components, 0, min(num_samples, num_features))
+
+    if Nx.rank(num_components) != 0 do
+      raise ArgumentError,
+            ":num_components must be a scalar tensor, got: #{inspect(num_components)}"
+    end
+
+    num_components =
+      cond do
+        Nx.Type.integer?(type) ->
+          clipped
+
+        Nx.Type.float?(type) ->
+          Nx.cumulative_sum(explained_variance)
+          |> then(&Nx.greater_equal(clipped, &1))
+          |> Nx.as_type({:s, 8})
+          |> Nx.argmax(tie_break: :low)
+          # add one to change an index into the number of clusters
+          |> Nx.add(1)
+
+        true ->
+          raise ArgumentError,
+                ":num_components must be an integer or float tensor, got: #{inspect(num_components)}"
+      end
+
+    %__MODULE__{
+      components: components,
+      explained_variance: explained_variance,
+      explained_variance_ratio: explained_variance_ratio,
+      singular_values: singular_values,
+      mean: mean,
+      num_components: num_components,
+      num_features: num_features,
+      num_samples: num_samples,
+      decomposer: decomposer
+    }
+  end
+
+  @doc """
+  For a fitted model performs a decomposition.
+
+  ## Options
+
+  #{NimbleOptions.docs(@transform_opts_schema)}
+
+  ## Returns
+
+  The function returns a decomposed data.
+  """
+  deftransform transform(model, opts \\ []) do
```

Comment:
Shouldn't we accept arbitrary dataset `x` and express it in terms of the trained principal components?

Reply:
Yes, I'll correct that.